src/c_api/c_api.cc

```diff
@@ -92,16 +95,589 @@ inline int MXAPIGetFunctionRegInfo(const FunRegType *e,
+    };
+
+    // stateful forward and backward
+    auto fstateful_lambda = [=](bool forward,
+                                const OpStatePtr& state_ptr,
+                                const OpContext& ctx,
+                                const std::vector<NDArray>& inputs,
+                                const std::vector<OpReqType>& req,
+                                const std::vector<NDArray>& outputs) {
+      // create a vector of tensors for inputs
+      std::vector<MXTensor> c_inputs(inputs.size());
+      for (size_t i = 0; i < inputs.size(); i++) {
+        c_inputs[i].data_ptr = inputs[i].data().dptr_;
+        c_inputs[i].dtype = (MXDType)inputs[i].dtype();
+        for (int_least16_t j = 0; j < inputs[i].shape().ndim(); j++) {
+          c_inputs[i].shape.push_back(inputs[i].shape().data()[j]);
+        }
+      }
+
+      // create a vector of tensors for outputs
+      std::vector<MXTensor> c_outputs(outputs.size());
+      for (size_t i = 0; i < outputs.size(); i++) {
+        c_outputs[i].data_ptr = outputs[i].data().dptr_;
+        c_outputs[i].dtype = (MXDType)outputs[i].dtype();
+        for (int j = 0; j < outputs[i].shape().ndim(); j++) {
+          c_outputs[i].shape.push_back(outputs[i].shape().data()[j]);
+        }
+      }
+
+      // get memory resource
+      const Resource &resource = ctx.requested[0];
+      mshadow::Stream<mxnet::cpu> *cpu_stream = ctx.get_stream<mxnet::cpu>();
+
+      // create lambda that captures stream & resource objects
+      auto cpu_alloc = [&](int size) {
+        mshadow::Tensor<mxnet::cpu, 1, char> data =
+        resource.get_space_typed<mxnet::cpu, 1, char>(mshadow::Shape1(size), cpu_stream);
+        return data.dptr_;
+      };
+
+      // create lambda without captures so that we can cast it to function pointer
+      // this needs to be a lambda function so that we can do the decltype cast
+      typedef decltype(cpu_alloc) alloc_type;
+      auto cpu_malloc = [](void* _cpu_alloc, int size) {
+        // cast the void* argument to the type for the cpu_alloc lambda function
+        alloc_type* cpualloc = static_cast<alloc_type*>(_cpu_alloc);
+        void* ptr = (*cpualloc)(size);
+        return ptr;
+      };
+
+      OpResource op_res(cpu_malloc, &cpu_alloc);
+
+      // retrieve op state object created from CreateOpState
+      CustomStatefulOpWrapper& op = state_ptr.get_state<CustomStatefulOpWrapper>();
+      CustomStatefulOp* state_op_inst = op.get_instance();
+      CHECK(state_op_inst != nullptr)
+      << "Error MXNet cannot load custom stateful operator'" << name_str << "'";
+
+      if (forward) {
+        CHECK(state_op_inst->Forward(c_inputs, c_outputs, op_res))
+        << "Error calling ForwardStateful for custom operator '" << name_str << "'";
+      } else {
+        CHECK(state_op_inst->Backward(c_inputs, c_outputs, op_res))
+        << "Error calling BackwardStateful for custom operator '" << name_str << "'";
```

Comment:
We cant call Forward/Backward directly, we need a _opCallStatefulFCompute function like _opCallFCompute that rebuilds the MXTensors in lib_api.h, not directly here in MXNet

Because we cannot pass vector objects between MXNet and the libraries. Vectors are a class, and we can only pass C-types (pointers, ints, etc.) for binary compatibility. This lets the library creator use a different version of gcc and/or cstdlib or STL, etc.